sets/sets.gobra

```diff
@@ -0,0 +1,529 @@
+ensures len(xs union ys) >= len(ys)
+decreases ys
+pure func UnionLenLower(xs, ys set[int]) utils.Unit {
+	return IsEmpty(ys) ? utils.Unit{} :
+		let y := choose(ys) in
+		(let yr := Remove(ys, y) in
+		(y in xs ?
+			(let xr := Remove(xs, y) in
+			(let _ := utils.Asserting(xr union yr == Remove(xs union ys, y)) in UnionLenLower(xr, yr))) :
+			(let _ := utils.Asserting(xs union yr == Remove(xs union ys, y)) in UnionLenLower(xs, yr))))
+}
+
+// The cardinality of a union of two sets is less than or equal to the cardinality of
+// both individual sets combined.
+ghost
+ensures len(xs union ys) <= len(xs) + len(ys)
+decreases
+pure func UnionLenUpper(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// The cardinality of the intersection of xs and ys is less than or equal to the
+// cardinality of xs.
+ghost
+ensures len(xs intersection ys) <= len(xs)
+decreases xs
+pure func IntersectLenUpper(xs, ys set[int]) utils.Unit {
+	return IsEmpty(xs) ? utils.Unit{} :
+		let x := choose(xs) in
+		(let _ := utils.Asserting((Remove(xs, x)) intersection ys == Remove((xs intersection ys), x)) in
+		(IntersectLenUpper(Remove(xs, x), ys)))
+}
+
+// The cardinality of the difference xs - ys is less than or equal to the cardinality of xs. 
+ghost
+ensures len(xs setminus ys) <= len(xs)
+decreases
+pure func DifferenceLenUpper(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+ensures len(xs union ys) == len(xs) + len(ys) - len(xs intersection ys)
+decreases
+pure func UnionLenEq(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+ensures len(xs setminus ys) == len(xs) - len(xs intersection ys)
+ensures len(xs setminus ys) + len(ys setminus xs) + len(xs intersection ys) == len(xs union ys)
+decreases
+pure func DifferenceLenEq(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+ensures (e in xs) ==> (len(Add(xs, e)) == len(xs))
+ensures !(e in xs) ==> (len(Add(xs, e)) == len(xs) + 1)
+decreases
+func AddLen(xs set[int], e int) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+ensures (e in xs) ==> (len(Remove(xs, e)) == len(xs) - 1)
+ensures !(e in xs) ==> (len(Remove(xs, e)) == len(xs))
+decreases
+func RemoveLen(xs set[int], e int) utils.Unit {
+	return utils.Unit{}
+}
+
+ghost
+requires AreDisjoint(xs, ys)
+ensures len(xs union ys) == len(xs) + len(ys)
+decreases
+pure func DisjointUnionLen(xs, ys set[int]) utils.Unit {
+	return utils.Unit{}
+}
+
+// QUES Do we want S1: LemmaBoundedSetSize? (Dafny proof requires LemmaSubsetSize +
+// contains unclear forall in proof)
+// TODO If we do want this; make sure to call it something like BoundedSetLen to be consistent.
+// Dafny signature:
+//   lemma LemmaBoundedSetSize(x: set<int>, a: int, b: int)
+//     requires forall i {:trigger i in x} :: i in x ==> a <= i < b
+//     requires a <= b
+//     ensures |x| <= b - a
+
+// *****************
+
+// QUES S3: min_elt{,_def}, max_elt{,_def} mention forall x . x in xs in
+// their post-condition; do we want these functions?
```

Comment:
what do these functions do?

Reply:
min_elt returns the smallest element in a non-empty set, while max_elt returns the largest element in a non-empty set.